src/validation.cpp

```diff
@@ -4356,7 +4356,18 @@ void CChainState::LoadExternalBlockFile(FILE* fileIn, FlatFilePos* dbp)
                     }
                 }
             } catch (const std::exception& e) {
-                LogPrintf("%s: Deserialize or I/O error - %s\n", __func__, e.what());
+                // historical bugs added extra data to the block files that does not deserialize cleanly.
+                // commonly this data is between readable blocks, but it does not really matter. such data is not fatal to the import process.
+                // the code that reads the block files deals with invalid data by simply ignoring it.
+                // it continues to search for the next {4 byte magic message start bytes + 4 byte length + block} that does deserialize cleanly
+                // and passes all of the other block validation checks dealing with POW and the merkle root, etc...
+                // we merely note with this informational log message when unexpected data is encountered.
+                // we could also be experiencing a storage system read error, or a read of a previous bad write. these are possible, but
+                // less likely scenarios. we don't have enough information to tell a difference here.
+                // the reindex process is not the place to attempt to clean and/or compact the block files. if so desired, a studious node operator
+                // may use knowledge of the fact that the block files are not entirely pristine in order to prepare a set of pristine, and
+                // perhaps ordered, block files for later reindexing.
+                LogPrint(BCLog::REINDEX, "%s: unexpected data at file offset 0x%x - %s. continuing\n", __func__, (nRewind - 1), e.what());
```

Comment:
I wonder whether it might be better to keep this an unconditional log, so that there is a higher chance of getting reports such as the one in #21379.
I think the expectation is that it shouldn't be possible to get this warning (unless there is actual disk corruption) after this PR, unless you have an existing datadir that is already slightly corrupted by past reindexes.

Reply:
I did consider keeping this log message unconditional. I think we'd get a fair number of false positive reports if we leave this scary log message as it was. This bug has been around for about 7 years, and maybe more if you count the bugs that the last fix tried to handle. So, there are probably a lot of blk files that have been _recoverably_ corrupted (i.e. corrupted in a way that is still processable). To me, it's more of a notification for a concerned node operator that really cares about having a clean set of blk files. The new message gives the exact offset to go look, and maybe edit the file, if you really want to. But, The existing message doesn't really do much to help the operator. It kind of just scares them, even though the processing goes on. Maybe there is a balance that we need to hit. But, I'm not sure what that looks like besides this right now.

Reply:
okay, makes sense. By the way, node operators also have the option to apply the `linearize-data.py` script in `/contrib/linearize` (which was adjusted in #16802 to deal with this bug, after a report in #14986), but this will also remove orphaned blocks from the blk files.

Reply:
The default loglevel is currently debug and will become info.

Warning and error levels are currently logged unconditionally, and the plan is to log info unconditionally as well.

I've made a note to check this in the severity logging updates.